### DISCUSSION TITLE
Auto-generated Table/Column Names

### DISCUSSION THREAD

**Discussion by daddz** [stephenafamo/bob#300] [Ideas]
Since there seem to be quite some great major changes already in the pipeline for the next version I thought it might be worthwhile to talk a bit about the auto-generated table and column names.

Currently (v0.28.1) there are multiple ways to get table and column names:
```
models.TableNames.MyTable -> string
models.MyTable.Name(ctx) -> Expression

models.ColumnNames.MyTable.MyColumn -> string
models.MyTableColumns.MyColumn -> Expression
models.MyTable.Columns() -> orm.Columns
```

A lot of the helper function accept `any` as their input which makes it a bit confusing on which variant can or needs to be used.

Are there any plans to streamline/simplify this?

I don't have a concrete idea to pitch just yet but wanted to see if there are already some ideas or plans before I dig into this more.

**Comment by stephenafamo** [comment]
I wish there is an easy way to merge these, but I can't think of one, and I can see them being used in different situations:

The methods that return an expression, should be preferred if building a query. Because this will always quote the table or column names and prevent issues in case the table or column name match an SQL keyword, or if it mixes upper and lower case.

The methods that return a string are primarily for situations where the user needs to maybe log something or just print the table name. There are also some situations that require an exact string, for example `ON CONFLICT (column_name)`.

`orm.Columns` is special because without it, it is quite tedious to build a list of properly qualified and quoted list of column names such as: `"user"."id" as "id", "user"."name" as "name", "user"."email" as "email"`.  
`orm.Columns` makes it easy to generate this list as an expression, even if the table name or column alias needs to be modified.
